File: app/src/java/java/com/example/ama/viewer/data/repo/DBRepositoryImpl.java
package com.example.ama.viewer.data.repo;

import com.example.ama.viewer.data.entity.GithubUser;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import io.realm.Realm;
import io.realm.RealmConfiguration;

public class DBRepositoryImpl implements DBRepository {

    private RealmConfiguration configuration;
    private Realm realm;

    public DBRepositoryImpl(RealmConfiguration configuration) {
        this.configuration = configuration;
    }

    @Override
    public Observable<GithubUser> getUserFromDb() {
        return Observable.defer(() -> Observable.fromCallable(() -> realm.where(GithubUser.class).findFirst()))
                .doOnSubscribe(disposable -> tryToOpenRealm())
                .doFinally(this::tryToCloseRealm)
                .subscribeOn(Schedulers.io());
    }

    @Override
    public Completable saveUserToDb(GithubUser user) {
        return Completable.fromAction(() -> save(user))
                .doOnSubscribe(disposable -> tryToOpenRealm())
                .doFinally(this::tryToCloseRealm)
                .subscribeOn(Schedulers.io());
    }

    private void save(GithubUser user) {
        realm.beginTransaction();
        realm.insertOrUpdate(user);
        realm.commitTransaction();
    }

    private void tryToOpenRealm() {
        if (realm == null) {
            realm = Realm.getInstance(configuration);
        }
    }

    private void tryToCloseRealm() {
        if (realm != null && !realm.isClosed()) {
            realm.close();
            realm = null;
        }
    }
}

File: app/src/kotlin/java/com/example/ama/viewer/data/repo/DBRepositoryImpl.kt
package com.example.ama.viewer.data.repo

import android.util.Log
import com.example.ama.viewer.data.entity.GithubUser
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import io.realm.Realm
import io.realm.RealmConfiguration

class DBRepositoryImpl(
        private val configuration: RealmConfiguration
) : DBRepository {

    var realm: Realm? = null

    override fun getUserFromDb(): Observable<GithubUser> =
            Observable.defer { Observable.fromCallable<GithubUser> { realm?.where(GithubUser::class.java)?.findFirst() } }
                    .doOnSubscribe { tryToOpenRealm() }
                    .doFinally { this.tryToCloseRealm() }
                    .subscribeOn(Schedulers.io())

    override fun saveUserToDb(user: GithubUser): Completable =
            Completable.fromAction { save(user) }
                    .doOnSubscribe { tryToOpenRealm() }
                    .doFinally { this.tryToCloseRealm() }
                    .subscribeOn(Schedulers.io())

    private fun save(user: GithubUser) {
        realm?.beginTransaction()
        realm?.insertOrUpdate(user)
        realm?.commitTransaction()
    }

    private fun tryToOpenRealm() {
        if (realm == null) {
            realm = Realm.getInstance(configuration)
        }
    }

    private fun tryToCloseRealm() {
        if (realm != null && !realm!!.isClosed) {
            realm!!.close()
            realm = null
        }
    }
}